File: solutions/scala/beer-song/1/src/main/scala/BeerSong.scala
object BeerSong {

  private val noMoreBottleVerse =
    "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"

  def recite(startBottleCount: Int, takeDown: Int): String = {
    (startBottleCount - takeDown + 1 to startBottleCount)
      .reverse
      .map {
             case x if x > 0 => recite(x)
             case _          =>
               noMoreBottleVerse
           }
      .mkString("\n")
  }

  private def recite(startBottleCount: Int) = {
    val firstLine  = reciteFirstLine(startBottleCount)
    val secondLine = reciteSecondLine(startBottleCount)
    s"$firstLine\n$secondLine\n"
  }

  private def reciteFirstLine(startBottleCount: Int): String = {
    val startBottleNoun = getBottleNoun(startBottleCount)
    s"$startBottleCount $startBottleNoun of beer on the wall, $startBottleCount $startBottleNoun of beer."
  }

  private def reciteSecondLine(startBottleCount: Int): String = {
    val remainingBottleCount = startBottleCount - 1
    if (remainingBottleCount > 0) {
      val remainingBottleNoun = getBottleNoun(remainingBottleCount)
      return s"Take one down and pass it around, $remainingBottleCount $remainingBottleNoun of beer on the wall."
    }
    "Take it down and pass it around, no more bottles of beer on the wall."
  }

  private def getBottleNoun(count: Int) =
    if (count == 1) {
      "bottle"
    }
    else {
      "bottles"
    }
}

File: solutions/csharp/beer-song/1/BeerSong.cs
using System.Linq;

public static class BeerSong
{
	private static readonly string NoMoreBottleVerse = """
			No more bottles of beer on the wall, no more bottles of beer.
			Go to the store and buy some more, 99 bottles of beer on the wall.
			""".Replace("\r", "");

	public static string Recite(int startBottleCount, int takeDown)
	{
		string[] verses = Enumerable.Range(startBottleCount - takeDown + 1, takeDown)
			.Reverse()
			.Select(x => x > 0
					? Recite(x)
					: NoMoreBottleVerse
			)
			.ToArray();

		return string.Join("\n\n", verses);
	}

	private static string Recite(int startBottleCount) =>
		$$"""
			{{ReciteFirstLine(startBottleCount)}}
			{{ReciteSecondLine(startBottleCount)}}
			"""
			.Replace("\r", "");

	private static string ReciteFirstLine(int startBottleCount)
	{
		string startBottleNoun = GetBottleNoun(startBottleCount);
		return $"{startBottleCount} {startBottleNoun} of beer on the wall, {startBottleCount} {startBottleNoun} of beer.";
	}

	private static string ReciteSecondLine(int startBottleCount)
	{
		int remainingBottleCount = startBottleCount - 1;
		if (remainingBottleCount > 0)
		{
			string remainingBottleNoun = GetBottleNoun(remainingBottleCount);
			return $"Take one down and pass it around, {remainingBottleCount} {remainingBottleNoun} of beer on the wall.";
		}
		else
		{
			return "Take it down and pass it around, no more bottles of beer on the wall.";
		}
	}

	private static string GetBottleNoun(int count) => (count == 1)
				? "bottle"
				: "bottles";
}